docs/dev/dev_environment.md
# Collecte-pro - Environnement de développement

## Environnement de développement

Collecte-pro est une application web Django, avec un front Vue.js. Elle utilise Postgres comme base de données et keycloak pour la gestion des utilisateurs.
Postgres et keycloak sont lancées via docker-compose, et le front est servi par le serveur Django en dev.

### Code source

Cloner le dépôt :

    git clone git@github.com:SocialGouv/collecte-pro.git collectepro

**Attention**, il faut avoir créer une clef SSH et l'avoir spécifiée sur github.

On se déplace dans le répertoire :

    cd collectepro

### Node

Installer node 24 et npm.

Installer les dependances node : `npm install`

Builder le front : `npm run build-all` (pour developper par la suite, on pourra utiliser
les commandes `watch` qui rebuildent au fur et à mesure des modifications. Voir
`package.json`)

### Variables d'environnement

Certaines variables d'environnement doivent être positionnées pour que l'application
fonctionne.

On définit les variables d'environnement dans le fichier `.env`.
On peut utiliser le fichier d'exemple comme ceci:

    cp .env.sample .env

Les variables d'environnement sont automatiquement intégrées au process uWSGI via le
fichier `ecc/wsgi.py` - de même pour le fichier `ecc/manage.py`.

Dans le fichier `.env`, modifier l'adresse de la db :

```
DATABASE_URL=postgres://ecc:ecc@localhost:5432/ecc
```

### Postgres et Keycloak

Lancer les conteneurs docker pour Postgres et Keycloak :

    docker-compose up -d

Keycloak est directement chargé avec un realm collecte-pro. Le compte administrateur de l’instance est «admin» avec le mot de passe «admin».
Pour créer les comptes administrateur collecte-pro ainsi que le compte administrateur du realm, if faut lancer le script `deploy/local/init_keycloak.sh` (il faut que le conteneur keycloak soit démarré).

    ./deploy/local/init_keycloak.sh

Pour charger les fonctions Postgres nécessaires pour les différents tableau de bord et les purge, il faut lancer le script `deploy/local/init_postgres.sh` (il faut que le conteneur postgres soit démarré).

    ./deploy/local/init_postgres.sh

### Python et Django

Le projet utilise Python 3.14. Cette version doit être installée sur la machine. Un venv est utilisé pour isoler les dépendances python.

    python3 -m venv venv
    source venv/bin/activate
    python3 -m pip install --upgrade pip
    pip install -r requirements.txt

Migrer la db : `python manage.py migrate`

Collecter les fichiers statiques : `python manage.py collectstatic --noinput`

Lancer le serveur local : `python manage.py runserver 0:8080`

Aller sur `http://localhost:8080/` et se logger avec le compte «admin_fonctionnel» et le mot de passe «1234».

## Création de nouvel utilisateur

### Création d’un nouvel utilisateur admin

Un nouvel utilisateur admin peut être créé via Keycloak, en se loggant sur l'interface d'administration de Keycloak avec le compte admin/admin.
Le nouvel utilisateur doit être créé dans le realm `collecte-pro`, et lui donner le rôle `admin` du client `client_1` pour qu'il puisse se logger sur l'interface d'administration de collecte-pro.

### Création d’un nouvel utilisateur répondant

Un nouveau répondant peut être créé via keycloak, en se loggant sur l'interface d'administration de Keycloak avec le compte admin/admin.
Aucun rôle n’est nécessaire pour se logger sur l’interface de collecte-pro, mais il faut que le compte soit activé et qu’il ait un mot de passe.

Il est aussi possible de créer un nouveau répondant via l’interface collecte-pro en l’ajoutant directement dans la liste des organisme interrogés.
Le compte sera créé dans keycloak et un mot de passe devras lui être ajouté.

## Restaurer/Sauvegarder la base de données en dev

Aucun dump n'est actuellement fourni par défaut car l'ancien était obsolète.

Le mot de passe est `ecc` (si créé comme signalé plus haut).

### Créer un nouveau dump

    pg_dump --verbose --clean --no-acl --no-owner -h postgres -U ecc -d ecc > db.dump

### Charger le dump dans la base

    psql -h localhost -U ecc -d ecc < db.dump

## Login et envoi d'emails

Les utlisateurs admin peuvent se logger à <http://localhost:8080/admin>.

Les utilisateurs non-admin doivent d'abord être créés via un utilisateur admin.

### Serveur d'email en local

Python contient un petit serveur SMTP, qui printe les mail dans la console au lieu de
les envoyer. C'est le plus simple pour developper.

Ajoutez les settings suivants dans `.env` :

```
export EMAIL_HOST='localhost'
export EMAIL_PORT=1025
export EMAIL_HOST_USER=''
export EMAIL_HOST_PASSWORD=''
export EMAIL_USE_TLS=False
export DEFAULT_FROM_EMAIL='testing@example.com'
```

Et lancez le serveur :
`python -m smtpd -n -c DebuggingServer localhost:1025`

## libmagic

Le serveur Django utilise libmagic (pour vérifier les types des fichiers uploadés), qui
doit être présent sur la machine. Vous pouvez essayer de démarrer sans, et si le serveur
lève une erreur c'est qu'il faut l'installer à la main sur votre machine.

Instructions d'installation données par django-magic, le package que nous utilisons :
<https://github.com/ahupp/python-magic#installation>

## Définition des locales

Cette plateforme utilise l'encodage UTF-8 à plusieurs endroit, notamment pour les nom de
fichiers uploadés.

Pour que cela fonctionne, il faut configurer correctement les 'locales', par exemple
comme ceci :

    localedef -c -f UTF-8 -i fr_FR fr_FR.UTF-8
    export LANG=fr_FR.UTF-8
    export LC_ALL=fr_FR.UTF-8

## Envoi d'emails périodiques

On utilise Celery Beat et Redis pour gérer l'envoi d'emails périodiques.

La fréquence des envois est configurée dans django admin, avec l'application
'django_celery_beat'.

Pour démarrer celery beat, il y a la commande suivante:

    celery worker --beat -A ecc -l info --scheduler django_celery_beat.schedulers:DatabaseScheduler &

Un autre façon de faire, est d'installer un service systemd:

    ln -s /opt/e-controle/deploy/conf/celery.service /etc/systemd/system/celery.service
    systemctl daemon-reload
    systemctl start celery
    systemctl restart status
    tail /var/log/ecc-celery.log

Si le serveur Redis n'est pas fourni, on peut l'installer:

    apt-get install redis
    systemctl start redis
    systemctl enable redis
    redis-cli ping

## uWSGI

Le server d'application uWSGI est utilisé en production.
Pour plus de détail : <https://uwsgi-docs.readthedocs.io/en/latest/>

## Parcel : Bundler JS

Nous avons fait le choix d'utiliser le bundler Parcel qui est une alternative à Webpack.
Voir le fichier ``package.json`` pour plus de détails.

Quelques commandes bash utiles:

    npm install  # Pour installer les dépendences

    npm run build-all

    npm run watch-control-detail  # Pour construire le fichier bundle en mode watch
    npm run build-control-detail  # Pour construire le fichier bundle

    npm run watch-questionnaire-create
    npm run watch-questionnaire-detail
    npm run watch-session-management
    npm run watch-sidebar

## Tests

### Backend tests

Lancer les tests :

    `pytest`

ou

    `pytest -s <dossier>`

(le flag -s sert a laisser le debugger prendre le controle si besoin).

### Frontend tests

Ils se situent dans `static/src/` avec le code, dans des dossiers `test`. Ce sont des
tests Vitest, pour trouver de la doc googler "test Vue with Vitest" par exemple.

Lancer les tests : `npm test`

Debugger un test : plusieurs debuggers sont possibles, dont Chrome Dev Tools et
Webstorm/Pycharm.

Vous pouvez également utiliser VSCode, voir la doc complète :
<https://code.visualstudio.com/docs/editor/debugging>


CHANGELOG.md
# Journal des modifications

Toutes les modifications notables de ce projet sont documentées dans ce fichier.

Ce format se base sur [keep a changelog](https://keepachangelog.com/fr/1.0.0/), et ce
projet adhère à [la Gestion sémantique de version](https://semver.org/lang/fr/spec/v2.0.0.html).

## [En cours]
## 4.1.0 - 2026-09-23
### Modifications
- CP-224 : Ajout d’une gauge pour la visualisation de l’espace de stockage utilisé par les fichiers déposés dans un espace de dépôt
## [4.0.0] - 2026-09-03
### Modifications
- CP-210 : Montée de version Django 4.2 vers 5.2.13 LTS
- CP-215 : Montée de version Python de 3.11 vers 3.14
## [3.4.8] - 2026-07-10
### Modifications
- CP-224 :  Téléchargement en masse via la fonction « Export ZIP » ne fonctionne plus
## [3.4.7] - 2026-07-08
### Modifications
- CP-223 :  Empêcher la saisie des caractères spéciaux pouvant poser problème dans l’application quand on ajoute un utilisateur (cas d'un demandeur ou d'un répondant déjà existant avec des caractères spéciaux non autorisés)
## [3.4.6] - 2026-07-03
### Modifications
- CP-223 :  Empêcher la saisie des caractères spéciaux pouvant poser problème dans l’application quand on ajoute un utilisateur
## [3.4.5] - 2026-03-30
### Modifications
- CP-218 :  Lien vers le site collecte-pro absent du message envoyé à un nouveau déposant
## [3.4.4] - 2026-03-24
### Modifications
- CP-216 :  Interdire la sélection de plusieurs fichiers à la fois
## [3.4.3] - 2026-02-27
### Modifications
- CP-211 :  Liste des utilisateurs en indiquant ceux avec le profil 'Inspecteur'
- CP-213 :  Test de V3.4.2 Anomalie sur l'export des fichiers (export vide)
## [3.4.2] - 2026-01-26
### Modifications
- CP-209 :  Montée de Django de 4.2.26 à 4.2.27
## [3.4.1] - 2026-01-15
### Modifications
- CP-204 :  RGAA :  Passer un scan de contrôle
- CP-205 :  Le Captcha ne fonctionne plus
## [3.4.0] - 2025-12-19
### Modifications
- CP-202 : Reprendre la montée de version Vue JS 3 pour mettre la stack technique à niveau
## [3.3.9] - 2025-11-13
### Modifications
- CP-197 : Correction de Vulnérabilité python-jose : confusion d’algorithme avec les clés ECDSA OpenSSH
- CP-198 : Problème Django : contournement de la validation lors de l’utilisation d’un champ pour uploader plusieurs fichiers
## [3.3.8] - 2025-10-20
### Modifications
- CP-199 : Adapations sur la cinématique de déconnexion liées à la migration Keycloak 16.1 -> 26.4
## [3.3.7] - 2025-09-19
### Modifications
- CP-195 : Purge des répondants "orphelins"
## [3.3.6] - 2025-07-09
### Modifications
- CP-191 : Mise en place d'un système de purge des espaces de dépôt (fix bug : tri par date la plus récente dans l’interface d’administration)
## [3.3.5] - 2025-07-04
### Modifications
- CP-193 : Rajouter une Date de création sur tous les objets (contrôle , contrôle question..) 
- CP-191 : Mise en place d'un système de purge des espaces de dépôt (tri par date la plus récente dans l’interface d’administration)
- Correction d'un bug sur le paramètre de purge
## [3.3.4] - 2025-06-04
### Modifications
- CP-191 : Mise en place d'un système de purge des espaces de dépôt (ajout de la date la plus récente dans l'interface d'administration)
- Correction d'un bug sur le paramètre de purge
## [3.3.3] - 2025-05-07
### Modifications
- CP-191 : Mise en place d'un système de purge des espaces de dépôt (cas des espaces de dépôt sans questionnaire)
- CP-194 : Rajouter une date de dernière modification au niveau Questionnaire (ajout de la date de dernière modification dans l'administration)
## [3.3.2] - 2025-04-10
### Modifications
- CP-191 : Mise en place d'un système de purge des espaces de dépôt
- CP-194 : Rajouter une date de dernière modification au niveau Questionnaire 
## [3.0.13] - 2025-03-27
### Modifications
- CP-196 : En production, les liens vers les modes d'emploi sont cassés
## [3.3.1] - 2025-03-14
### Modifications
- CP-189 : Le choix d'un modèle d'espace de dépôt doit être indiqué comme non obligatoire à la création
- CP-192 : Ajout de la requête CP-190 dans le tableau de bord
## [3.3.0] - 2025-03-12
### Modifications
- CP-187 : La duplication d'un questionnaire crée un modèle
## [3.2.1] - 2025-01-27
### Modifications
- CP-184 : Vulnérabilités Collecte-Pro
## [3.2.0] - 2025-01-10
### Modifications
- CP-185 : Possiblité de sélectionner un modèle lors de la création d'un espace de dépôt
## [3.1.1] - 2024-12-30
### Modifications
- CP-182 : Dans la liste des espaces de dépôt, identifier les modèles
- CP-183 : Possibilité d'épingler en haut de la liste des dépôts, les espaces de dépôt modèles
## [3.1.0] - 2024-12-09
### Modifications
- CP-179 : Identifier un espace de dépôt comme modèle
- CP-180 : Identifier un espace de dépôt comme modèle via l'accès rapide
- CP-181 : Identifier un espace de dépôt comme modèle quand celui-ci est dupliqué
- CP-68 : Contre Audit RGAA V2
## [3.0.12] - 2024-11-20
### Modifications
- CP-20 : Identifier les espaces modèle - Lot 1
## [3.0.11] - 2024-11-13
### Modifications
- CP-109 : Anomalie d'adresse email
## [3.0.10] - 2024-11-08
### Modifications
- CP-68 : Contre Audit RGAA V1
## [3.0.9] - 2024-09-27
### Modifications
- CP-96 : Modifier la phrase d'exemple au dessus du champ de présentation du questionnaire
- CP-101 : Pages communication : ouvrir le lien vers l'instance de démo dans une nouvelle fenêtre
- CP-109 : Instance Démo : publier les bons codes de connexion
## [3.0.8] - 2024-07-08
### Modifications
- CP-165 : La fonctionnalité Action renvoie un code erreur
## [3.0.7] - 2024-07-04
### Modifications
- CP-158 : Problème d'accès à l'espace de démo
- CP-145 : Livraison de Captcha V1
## [3.0.6] - 2024-06-17
### Modifications
- CP-168 : Le message d'alerte n'apparaît plus sur l'application
## [3.0.5] - 2024-06-14
### Modifications
- CP-167 : Fonction d'export Excel des fichiers par le répondant KO
## [3.0.4] - 2024-06-05
### Modifications
- CP-145 : Doublon de mails
## [3.0.3] - 2024-05-30
### Modifications
- CP-145 : Mettre un captcha sur le formulaire de contact 

ecc/settings.py
import locale
import os
import environ
from dotenv import load_dotenv

env = environ.Env(
    DEBUG=(bool, False),
    EMAIL_USE_TLS=(bool, False),
    EMAIL_USE_SSL=(bool, False),
)

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/3.2/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = os.environ.get('SECRET_KEY')

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = env('DEBUG')

ALLOWED_HOSTS = ['*']

SHOW_DEBUG_TOOLBAR = env('SHOW_DEBUG_TOOLBAR', default=DEBUG)

DEBUG_TOOLBAR_CONFIG = {
    "SHOW_TOOLBAR_CALLBACK": lambda r: SHOW_DEBUG_TOOLBAR,
    'SHOW_COLLAPSED': True,
}

# Application definition

INSTALLED_APPS = [
    # Django apps
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'whitenoise.runserver_nostatic',
    'django.contrib.staticfiles',
    'django.contrib.sites',
    'mozilla_django_oidc',  # Load after auth

    # Third-party apps
    'debug_toolbar',
    'model_utils',
    'ordered_model',
    'django_extensions',
    'actstream',
    'rest_framework',
    'drf_spectacular',
    'celery',
    'django_celery_beat',
    'django_cleanup.apps.CleanupConfig',
    'django_ckeditor_5',
    'django_filters',
    'email_obfuscator',
    'django_softdelete',

    # Project's apps
    'backoffice',
    'config',
    'control',
    'demo',
    'editor',
    'faq',
    'reporting',
    'user_profiles',
    'utils',
    'session',
    'soft_deletion',
    'tos',
    'logs',
    'parametres',
    'alerte',
    'presentation',

    # Central app - loaded last
    'ecc',
]

# Keycloak configuration
KEYCLOAK_ACTIVE = env('KEYCLOAK_ACTIVE', default=False)
if KEYCLOAK_ACTIVE:
    KEYCLOAK_URL = env('KEYCLOAK_URL', default='http://localhost:8080/')
    KEYCLOAK_REALM = env('KEYCLOAK_REALM', default='collectepro')
    OPENID_PREFIX = f'{KEYCLOAK_URL}realms/{KEYCLOAK_REALM}/protocol/openid-connect'
    OIDC_OP_JWKS_ENDPOINT = f'{OPENID_PREFIX}/certs'
    OIDC_OP_AUTHORIZATION_ENDPOINT = f'{OPENID_PREFIX}/auth'
    OIDC_OP_TOKEN_ENDPOINT = f'{OPENID_PREFIX}/token'
    OIDC_OP_USER_ENDPOINT = f'{OPENID_PREFIX}/userinfo'
    OIDC_OP_LOGOUT_ENDPOINT = f'{OPENID_PREFIX}/logout'
    OIDC_OP_LOGOUT_URL_METHOD = env('OIDC_OP_LOGOUT_URL_METHOD', default='ecc.provider_logout')
    OIDC_RP_CLIENT_ID = env('OIDC_RP_CLIENT_ID', default='')
    OIDC_RP_CLIENT_SECRET = env('OIDC_RP_CLIENT_SECRET', default='')
    OIDC_VERIFY_SSL = env('OIDC_VERIFY_SSL', default=True)
    OIDC_RP_SIGN_ALGO = env('OIDC_RP_SIGN_ALGO', default='RS256')
    OIDC_STORE_ID_TOKEN = True
    AUTHENTICATION_BACKENDS = (
        'ecc.auth.EccOIDCAuthenticationBackend',
    )
    KEYCLOAK_ADMIN_USERNAME = env('KEYCLOAK_ADMIN_USERNAME', default='admin')
    KEYCLOAK_ADMIN_PASSWORD = env('KEYCLOAK_ADMIN_PASSWORD', default='admin')
    KEYCLOAK_URL_CLIENT_ID = env('KEYCLOAK_URL_CLIENT_ID', default='url_client_id')

LOGIN_REDIRECT_URL = "/accueil"
LOGOUT_REDIRECT_URL = "/"

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django_permissions_policy.PermissionsPolicyMiddleware',
    'csp.middleware.CSPMiddleware',
    'whitenoise.middleware.WhiteNoiseMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django_session_timeout.middleware.SessionTimeoutMiddleware',
    'debug_toolbar.middleware.DebugToolbarMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'tos.middleware.WelcomeMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.contrib.sites.middleware.CurrentSiteMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'ecc.custom_cache_control_middleware.CustomCacheControlMiddleware',
]

ROOT_URLCONF = 'ecc.urls'

TEMPLATE_DIR = os.path.join(BASE_DIR, 'templates')

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [TEMPLATE_DIR],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'ecc.context_processors.current_site',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
                'django_settings_export.settings_export',
            ],
        },
    },
]

WSGI_APPLICATION = 'ecc.wsgi.application'


# Database
# https://docs.djangoproject.com/en/3.2/ref/settings/#databases

DATABASES = {'default': env.db()}

SITE_ID = 1

# Password validation
# https://docs.djangoproject.com/en/3.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]

# HTTP Security
SECURE_CONTENT_TYPE_NOSNIFF = True
SECURE_REFERRER_POLICY = 'same-origin'
FEATURE_POLICY = {
    'geolocation': 'none',
    'autoplay': ['self', ],
    'accelerometer': 'none',
    'camera': 'none',
    'gyroscope': 'none',
    'magnetometer': 'none',
    'microphone': 'none',
    'payment': 'none',
    'usb': 'none',
}
# Strict-Transport-Security
SECURE_BROWSER_XSS_FILTER = True
CSRF_COOKIE_SECURE = True
CSRF_COOKIE_SAMESITE = 'Strict'
SECURE_SSL_REDIRECT = True
SESSION_COOKIE_SECURE = True
SESSION_COOKIE_HTTPONLY = True
# Can't use Strict mode in order to use Keycloak
SESSION_COOKIE_SAMESITE = 'Lax'
X_FRAME_OPTIONS = 'DENY'
SECURE_HSTS_SECONDS = 30
# Content-Security-Policy
CSP_DEFAULT_SRC = env('CSP_DEFAULT_SRC', default=("'self'", "'frame-src youtube.com www.youtube.com;'"))

CSP_STYLE_SRC = env('CSP_STYLE_SRC', default=("'self'", "'unsafe-inline'"))
CSP_STYLE_SRC += ("https://static.piste.gouv.fr",)

CSP_SCRIPT_SRC = env('CSP_SCRIPT_SRC', default=("'self'", "'unsafe-eval'", "'unsafe-inline'"))
CSP_SCRIPT_SRC += ("https://code.jquery.com",)

CSP_IMG_SRC = env('CSP_IMG_SRC', default=("'self'", "https:",))

if DEBUG:
    CSRF_COOKIE_SECURE = False
    SECURE_SSL_REDIRECT = False
    SESSION_COOKIE_SECURE = False

# Email
EMAIL_BACKEND = env('EMAIL_BACKEND', default='django.core.mail.backends.smtp.EmailBackend')
EMAIL_HOST = env('EMAIL_HOST')
EMAIL_PORT = env('EMAIL_PORT')
EMAIL_HOST_USER = env('EMAIL_HOST_USER',)
EMAIL_HOST_PASSWORD = env('EMAIL_HOST_PASSWORD')
EMAIL_TIMEOUT = env('EMAIL_TIMEOUT', default=3)

EMAIL_USE_TLS = env('EMAIL_USE_TLS')
EMAIL_USE_SSL = env('EMAIL_USE_SSL')

# Time we wait in between emails, to space them out and avoid going over our allowed email quota
EMAIL_SPACING_TIME_MILLIS = env('EMAIL_SPACING_TIME_MILLIS', default=10000)

# The user will get a warning when trying to add an inspector whose email doesn't end with
# EXPECTED_INSPECTOR_EMAIL_ENDINGS
EXPECTED_INSPECTOR_EMAIL_ENDINGS = env('EXPECTED_INSPECTOR_EMAIL_ENDINGS', default='')

SEND_EMAIL_WHEN_USER_ADDED = env('SEND_EMAIL_WHEN_USER_ADDED', default=False)
SEND_EMAIL_WHEN_USER_REMOVED = env('SEND_EMAIL_WHEN_USER_REMOVED', default=False)

DEFAULT_FROM_EMAIL = env('DEFAULT_FROM_EMAIL')

LOGIN_URL = 'login'


# Internationalization
# https://docs.djangoproject.com/en/3.2/topics/i18n/

LANGUAGE_CODE = 'fr-fr'
TIME_ZONE = 'Europe/Paris'
USE_I18N = True
USE_L10N = True
USE_TZ = True

# A trick for DRF that does not seems to know about the locale
try:
    locale.setlocale(locale.LC_TIME, "fr_FR.UTF-8")
except locale.Error:
    pass  # setlocale can crash, for instance when running on Heroku.

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/3.2/howto/static-files/

ADMIN_URL = env('ADMIN_URL', default='admin/')

# Exclude incoming file if its mime type contains any of the following text
UPLOAD_FILE_MIME_TYPE_BLACKLIST = env(
    'UPLOAD_FILE_MIME_TYPE_BLACKLIST',
    default=('exe', 'msi', 'script'))

UPLOAD_FILE_EXTENSION_BLACKLIST = env('UPLOAD_FILE_EXTENSION_BLACKLIST', default=('.exe', '.dll', '.sql'))

UPLOAD_FILE_MAX_SIZE_MB = env('UPLOAD_FILE_MAX_SIZE_MB', default=256)

MAX_FILENAME_LENGTH = env('MAX_FILENAME_LENGTH', default=150)

STATIC_URL = '/static/'

# Collect static won't work if you haven't configured this
# django.core.exceptions.ImproperlyConfigured: You're using the staticfiles app without having set
#  the STATIC_ROOT setting to a filesystem path.
DEFAULT_STATIC_ROOT = os.path.join(BASE_DIR, 'staticfiles')
STATIC_ROOT = env('STATIC_ROOT', default=DEFAULT_STATIC_ROOT)

STATICFILES_DIRS = [
    os.path.join(BASE_DIR, "static"),
    # NPM modules that we need to link in <script> tags directly : add them to static files.
    os.path.join(BASE_DIR, 'node_modules', 'bootstrap', 'dist', 'js'),
    os.path.join(BASE_DIR, 'node_modules', 'jquery', 'dist'),
]

# Want forever-cacheable files and compression support?
# STATICFILES_STORAGE = 'whitenoise.storage.CompressedManifestStaticFilesStorage'

MEDIA_URL = '/media/'
DEFAULT_MEDIA_ROOT = os.path.join(BASE_DIR, 'media')
MEDIA_ROOT = env('MEDIA_ROOT', default=DEFAULT_MEDIA_ROOT)

SENDFILE_BACKEND = env('SENDFILE_BACKEND', default='sendfile.backends.simple')

PIWIK_TRACKER_BASE_URL = env('PIWIK_TRACKER_BASE_URL', default=None)
PIWIK_SITE_ID = env('PIWIK_SITE_ID', default=None)

SETTINGS_EXPORT = [
    'PIWIK_SITE_ID',
    'PIWIK_TRACKER_BASE_URL',
    'SESSION_EXPIRE_SECONDS',
    'DEBUG',
    'SAVE_IP_ADDRESS',
    'ENV_NAME',
    'KEYCLOAK_ACTIVE',
]

REST_FRAMEWORK = {
    'DEFAULT_RENDERER_CLASSES': [
        'rest_framework.renderers.JSONRenderer',
    ],
    'DEFAULT_PERMISSION_CLASSES': (
        'rest_framework.permissions.IsAuthenticated',
    ),
    'DEFAULT_FILTER_BACKENDS': (
        'django_filters.rest_framework.DjangoFilterBackend',
        'rest_framework.filters.SearchFilter',
    ),
    'DEFAULT_SCHEMA_CLASS': 'drf_spectacular.openapi.AutoSchema',
}

if DEBUG:
    REST_FRAMEWORK['DEFAULT_RENDERER_CLASSES'].append(
        'rest_framework.renderers.BrowsableAPIRenderer',
    )

CELERY_BROKER_URL = env('CELERY_BROKER_URL')
CELERY_QUEUE = env('CELERY_QUEUE', default='default')
HTTP_AUTHORIZATION = env('HTTP_AUTHORIZATION', default=None)

CKEDITOR_5_CONFIGS = {
    'default': {
        'toolbar': {
            'items': [
                'bold', 'italic', 'underline',
                '|', 'numberedList', 'bulletedList',
                '|', 'link',
                '|', 'removeFormat', 'sourceEditing',
            ],
        },
    },
}

CKEDITOR_5_FILE_UPLOAD_PERMISSION = "staff"

# Demo application
DEMO_INSPECTOR_USERNAME = env('DEMO_INSPECTOR_USERNAME', default=None)
DEMO_AUDITED_USERNAME = env('DEMO_AUDITED_USERNAME', default=None)
ALLOW_DEMO_LOGIN = env('ALLOW_DEMO_LOGIN', default=False)

# Session management
SESSION_EXPIRE_SECONDS = env('SESSION_EXPIRE_SECONDS', default=24 * 60 * 60)
SESSION_EXPIRE_AFTER_LAST_ACTIVITY = True

# Ip adress
SAVE_IP_ADDRESS = env('SAVE_IP_ADDRESS', default=False)

# Environnement name
ENV_NAME = env('ENV_NAME', default='')

# Url of collecte-pro in questionnaire
QUESTIONNAIRE_SITE_URL = env('QUESTIONNAIRE_SITE_URL', default='')

# Mise à jour pour Django>=3.2
DEFAULT_AUTO_FIELD = 'django.db.models.AutoField'

# Jours de relance par défaut avant échéance d'un questionnaire
JOURS_ECHEANCE = env('JOURS_ECHEANCE', default=7)

# Indique si la page par défaut est celle de présentation
PRESENTATION_ACTIVE = env('PRESENTATION_ACTIVE', default=True)

# Captcha
load_dotenv()
GRANT_TYPE = os.getenv('GRANT_TYPE')
CLIENT_ID = os.getenv('CLIENT_ID')
CLIENT_SECRET = os.getenv('CLIENT_SECRET')
SCOPE = os.getenv('SCOPE')

OAUTH_URL = os.getenv('OAUTH_URL')
SIMPLE_CAPTCHA_ENDPOINT_URL = os.getenv('SIMPLE_CAPTCHA_ENDPOINT_URL')
VALIDER_CAPTCHA_URL = os.getenv('VALIDER_CAPTCHA_URL')
